### PR TITLE
type: fix type error

### LINKED ISSUE #41643: Type definition of modal buttons doesn't allow `data-*` attributes
### Reproduction link

[Edit on CodeSandbox](https://codesandbox.io/s/antd-typescript-forked-c9bbi8?file=/demo.tsx)

### Steps to reproduce

Try to pass any `data-*` attribute to the cancel or the submit buttons of the modal. This will result in a TS error as it's not allowed. Adding `@ts-ignore` comment will show that will actually work

### What is expected?

No TS error

### What is actually happening?

Error stating that
```
Object literal may only specify known properties, and '"data-testid"' does not exist in type
```

| Environment | Info |
| --- | --- |
| antd | 5.3.3 |
| React | Irrelevant |
| System | Irrelevant |
| Browser | Irrelevant |

### PR DESCRIPTION
**Pull request by li-jia-nan**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [x] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

- close #41643

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | type: fix type error |
| 🇨🇳 Chinese | type: fix type error |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at 516bdae</samp>

Added custom data attributes and type casting to `button.tsx`. This allows users to pass arbitrary data to buttons and prevents type errors.

### <samp>🤖 Generated by Copilot at 516bdae</samp>

* Extend `ButtonProps` type to support custom data attributes ([link](https://github.com/ant-design/ant-design/pull/41650/files?diff=unified&w=0#diff-463cd38ab465e1f3b90463a403d8e8399f926361436eaa9adb8a76a8bf9d4140L57-R61))
* Reorder import statements to group type imports together ([link](https://github.com/ant-design/ant-design/pull/41650/files?diff=unified&w=0#diff-463cd38ab465e1f3b90463a403d8e8399f926361436eaa9adb8a76a8bf9d4140L5-R17))
* Cast `rest` props to `ButtonProps` instead of `AnchorButtonProps` ([link](https://github.com/ant-design/ant-design/pull/41650/files?diff=unified&w=0#diff-463cd38ab465e1f3b90463a403d8e8399f926361436eaa9adb8a76a8bf9d4140L204-R207))

### CODE REVIEW

**Comment by li-jia-nan** [comment]
- https://github.com/microsoft/TypeScript/issues/32447

找 @RexSkz  大佬请教了一下，根据上面这个 issue，在 tsx 里面写带有横杠的属性，都不会做类型检查，所以其实 `data-*` 和 `aria-*` 压根就没在类型定义里出现过，而是 tsc 跳过了类型检查，这个现象在组件里也是如此，并不是原生 DOM 元素特有的，这也就意味着，带横杠的属性无法通过 `{ ...obj }` 透传进去，必须直接写在 jsx tag 里面，虽然现在可以通过 template literal type 实现类型校验（issue 的后面有人提了），但是不太建议在 antd 里面这么干

File: components/button/button.tsx [hunk 1]
```diff
@@ -54,7 +53,11 @@ export type NativeButtonProps = {
 } & BaseButtonProps &
   Omit<React.ButtonHTMLAttributes<HTMLButtonElement>, 'type' | 'onClick'>;
 
-export type ButtonProps = Partial<AnchorButtonProps & NativeButtonProps>;
+interface Dataset {
+  [key: `data-${string}`]: string;
+}
+
+export type ButtonProps = Partial<AnchorButtonProps & NativeButtonProps & Dataset>;
```
Comment (MadCcc) [inline components/button/button.tsx:60]: 直接加到 BaseButtonProps 里？

**Comment by MadCcc** [comment]
> 这也就意味着，带横杠的属性无法通过 `{ ...obj }` 透传进去，必须直接写在 jsx tag 里面

类型校验和 spread 有关系么，这个就算被 tsc 跳过了编译之后应该是会透穿进去吧，毕竟 antd 不是用 tsc 编译的

**Comment by li-jia-nan** [comment]
> 类型校验和 spread 有关系么，这个就算被 tsc 跳过了编译之后应该是会透穿进去吧，毕竟 antd 不是用 tsc 编译的

和 spread 无关，这里我表达有误，是会被透穿进去的，我想表达的是在对象中没有这个 key 所以会报错，所以最好直接写在标签上

File: components/button/button.tsx (hunk 1 above)
Comment (li-jia-nan) [reply components/button/button.tsx:60]: done~

**Review by MadCcc — APPROVED** [review]

**Pull request merged**